js/apps/account-ui/src/personal-info/PersonalInfo.tsx

```diff
@@ -6,87 +6,104 @@ import {
-  const form = useForm<UserRepresentation>({ mode: "onChange" });
+  const [supportedLocales, setSupportedLocales] = useState<string[]>([]);
+  const form = useForm<UserFormFields>({ mode: "onChange" });
   const { handleSubmit, reset, setError } = form;
   const { addAlert, addError } = useAlerts();
 
   usePromise(
-    (signal) => getPersonalInfo({ signal }),
-    (personalInfo) => {
+    (signal) =>
+      Promise.all([
+        getPersonalInfo({ signal }),
+        getSupportedLocales({ signal }),
+      ]),
+    ([personalInfo, supportedLocales]) => {
       setUserProfileMetadata(personalInfo.userProfileMetadata);
+      setSupportedLocales(supportedLocales);
       reset(personalInfo);
     },
   );
 
-  const onSubmit = async (user: UserRepresentation) => {
+  const onSubmit = async (user: UserFormFields) => {
     try {
       await savePersonalInfo(user);
       keycloak?.updateToken();
       addAlert(t("accountUpdatedMessage"));
     } catch (error) {
       addError(t("accountUpdatedError").toString());
 
-      (error as FieldError[]).forEach((e) => {
-        const params = Object.assign(
-          {},
-          e.params.map((p) => t((isBundleKey(p) ? unWrap(p) : p) as TFuncKey)),
-        );
-        setError(fieldName(e.field) as keyof UserRepresentation, {
-          message: t(e.errorMessage as TFuncKey, {
-            ...params,
-            defaultValue: e.field,
-          }),
-          type: "server",
-        });
-      });
+      setUserProfileServerError(
+        error,
+        setError,
+        (key: TFuncKey, param?: object) => t(key, { ...param }),
+      );
     }
   };
 
   if (!userProfileMetadata) {
     return <Spinner />;
   }
 
+  const {
+    updateEmailFeatureEnabled,
+    updateEmailActionEnabled,
+    isRegistrationEmailAsUsername,
+    isEditUserNameAllowed,
+  } = environment.features;
   return (
     <Page title={t("personalInfo")} description={t("personalInfoDescription")}>
       <Form isHorizontal onSubmit={handleSubmit(onSubmit)}>
-        <FormProvider {...form}>
-          <UserProfileFields metaData={userProfileMetadata} />
-        </FormProvider>
+        <UserProfileFields
+          form={form}
+          userProfileMetadata={userProfileMetadata}
+          supportedLocales={supportedLocales}
+          t={(key: TFuncKey) => t(key)}
+          renderer={(attribute) =>
+            attribute.name === "email" &&
+            updateEmailFeatureEnabled &&
+            updateEmailActionEnabled &&
+            (!isRegistrationEmailAsUsername || isEditUserNameAllowed) ? (
+              <Button
+                id="update-email-btn"
+                variant="link"
+                onClick={() =>
+                  keycloak?.keycloak.login({ action: "UPDATE_EMAIL" })
+                }
+                icon={<ExternalLinkSquareAltIcon />}
+                iconPosition="right"
+              >
+                {t("updateEmail")}
+              </Button>
+            ) : undefined
```

Comment:
I think this could use an AND operator here since the else condition doesn't render anything.

```tsx
(!isRegistrationEmailAsUsername || isEditUserNameAllowed) && (
  <Button></Button>
)
```